### PR TITLE
Release `v4.0.0-alpha.1`

### PR DESCRIPTION
**Pull request by HCastano**
Two main reasons I want to get this out.

1. It's the beginning of some changes related to the metadata, and we need an ink!
   release in order to make `cargo-contract` releases which can understand this.
2. Green's storage PR is going to get merged soon, and I'd prefer if that was in a
   standalone release.

### CODE REVIEW

**Comment by HCastano** [comment]
Hmm, will look into CI failures tomorrow 🕵️

File: CHANGELOG.md [hunk 1]
```diff
@@ -6,16 +6,26 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
-- Add Mapping::contains(key) and Mapping::insert_return_size(key, val) ‒ [#1224](https://github.com/paritytech/ink/pull/1224)
-- Optimise deny_payment. Use everywhere semantic of deny ‒ [#1267](https://github.com/paritytech/ink/pull/1267)
-- Implement ecdsa_to_eth_address() and remove eth_compatibility crate ‒ [#1233](https://github.com/paritytech/ink/pull/1233)
+## Breaking Changes
+This release contains a few breaking changes. These are indicated with the :x: emoji.
+Most of these were intitially introduced in `v3.1.0` and `v3.2.0` releases but
+compatibility was restored in `v3.3.0`.
+
+## Added
+- :x: Add Mapping::contains(key) and Mapping::insert_return_size(key, val) ‒ [#1224](https://github.com/paritytech/ink/pull/1224)
```
Comment (cmichi) [inline CHANGELOG.md:17]:
```suggestion
- :x: Add `Mapping::contains(key)` and `Mapping::insert_return_size(key, val)` ‒ [#1224](https://github.com/paritytech/ink/pull/1224)
```

File: CHANGELOG.md [hunk 2]
```diff
@@ -6,16 +6,26 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+- Add payment-channel example - [#1248](https://github.com/paritytech/ink/pull/1248)
+- :x: Add `version` field to ink! metadata - [#1313](https://github.com/paritytech/ink/pull/1313)
```
Comment (cmichi) [inline CHANGELOG.md:19]:
```suggestion
- :x: Add `Mapping::contains(key)` and `Mapping::insert_return_size(key, val)` ‒ [#1224](https://github.com/paritytech/ink/pull/1224)
- Add [`payment-channel`](https://github.com/paritytech/ink/tree/master/examples/payment-channel) example ‒ [#1248](https://github.com/paritytech/ink/pull/1248)
- :x: Add `version` field to ink! metadata ‒ [#1313](https://github.com/paritytech/ink/pull/1313)
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -6,16 +6,26 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 ## [Unreleased]
 
-There are a number of backwards incompatible changes which are on the `master` branch
-waiting to be released. These are:
+# Version 4.0.0-alpha.1
```
Comment (cmichi) [inline CHANGELOG.md:9]:
```suggestion
## Version 4.0.0-alpha.1
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -6,16 +6,26 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
 
-- Add Mapping::contains(key) and Mapping::insert_return_size(key, val) ‒ [#1224](https://github.com/paritytech/ink/pull/1224)
-- Optimise deny_payment. Use everywhere semantic of deny ‒ [#1267](https://github.com/paritytech/ink/pull/1267)
-- Implement ecdsa_to_eth_address() and remove eth_compatibility crate ‒ [#1233](https://github.com/paritytech/ink/pull/1233)
+## Breaking Changes
```
Comment (cmichi) [inline CHANGELOG.md:11]:
```suggestion
### Breaking Changes
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -6,16 +6,26 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+This release contains a few breaking changes. These are indicated with the :x: emoji.
+Most of these were intitially introduced in `v3.1.0` and `v3.2.0` releases but
+compatibility was restored in `v3.3.0`.
+
+## Added
```
Comment (cmichi) [inline CHANGELOG.md:16]:
```suggestion
### Added
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -6,16 +6,26 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+- :x: Add Mapping::contains(key) and Mapping::insert_return_size(key, val) ‒ [#1224](https://github.com/paritytech/ink/pull/1224)
+- Add payment-channel example - [#1248](https://github.com/paritytech/ink/pull/1248)
+- :x: Add `version` field to ink! metadata - [#1313](https://github.com/paritytech/ink/pull/1313)
 - The `rand-extension` example has been adapted to an updated version of the `ChainExtension` API ‒ [#1356](https://github.com/paritytech/ink/pull/1356)
 
-### Compatibility
+## Changed
```
Comment (cmichi) [inline CHANGELOG.md:22]:
```suggestion
### Changed
```

File: CHANGELOG.md [hunk 7]
```diff
@@ -6,16 +6,26 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+- :x: Contract size optimization in case contract doesn't accept payment ‒ [#1267](https://github.com/paritytech/ink/pull/1267) (thanks [@xgreenx](https://github.com/xgreenx)).
+
+## Removed
```
Comment (cmichi) [inline CHANGELOG.md:25]:
```suggestion
### Removed
```

File: CHANGELOG.md [hunk 8]
```diff
@@ -6,16 +6,26 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+- :x: Implement ecdsa_to_eth_address() and remove eth_compatibility crate ‒ [#1233](https://github.com/paritytech/ink/pull/1233)
 
+## Compatibility
```
Comment (cmichi) [inline CHANGELOG.md:28]:
```suggestion
### Compatibility
```

File: CHANGELOG.md [hunk 9]
```diff
@@ -6,16 +6,26 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
-waiting to be released. These are:
+# Version 4.0.0-alpha.1
 
-- Add Mapping::contains(key) and Mapping::insert_return_size(key, val) ‒ [#1224](https://github.com/paritytech/ink/pull/1224)
-- Optimise deny_payment. Use everywhere semantic of deny ‒ [#1267](https://github.com/paritytech/ink/pull/1267)
-- Implement ecdsa_to_eth_address() and remove eth_compatibility crate ‒ [#1233](https://github.com/paritytech/ink/pull/1233)
+## Breaking Changes
+This release contains a few breaking changes. These are indicated with the :x: emoji.
+Most of these were intitially introduced in `v3.1.0` and `v3.2.0` releases but
+compatibility was restored in `v3.3.0`.
```
Comment (cmichi) [inline CHANGELOG.md:14]: Could you add something on how to migrate and what the implications are? So along the lines of https://github.com/paritytech/ink/blob/master/CHANGELOG.md#breaking-changes.

**Review by cmichi — CHANGES_REQUESTED** [review]

File: CHANGELOG.md [hunk 10]
```diff
@@ -6,16 +6,36 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+  `ecdsa_to_eth_address()` function from it can now be found in the `ink_env` crate.
+- As part of [#1267](https://github.com/paritytech/ink/pull/1267) an argument to `ink_lang::codegen::execute_constructor()` (which is
+  used internally by the ink! macros) was removed.
+- As part of [#1313](https://github.com/paritytech/ink/pull/1313) the ink! ABI was changed so that the version was specified using a
+  dedicated `version` key instead of an implicit key which wrapped the entire ABI.
+
+### Added
+- :x: Add `Mapping::contains(key)` and `Mapping::insert_return_size(key, val)` ‒ [#1224](https://github.com/paritytech/ink/pull/1224)
+- Add [`payment-channel`](https://github.com/paritytech/ink/tree/master/examples/payment-channel) example - [#1248](https://github.com/paritytech/ink/pull/1248)
+- :x: Add `version` field to ink! metadata - [#1313](https://github.com/paritytech/ink/pull/1313)
```
Comment (cmichi) [inline CHANGELOG.md:28]:
```suggestion
- Add [`payment-channel`](https://github.com/paritytech/ink/tree/master/examples/payment-channel) example ‒ [#1248](https://github.com/paritytech/ink/pull/1248)
- :x: Add `version` field to ink! metadata ‒ [#1313](https://github.com/paritytech/ink/pull/1313)
```

screenshot-github com-2022 08 24-18_52_01

File: CHANGELOG.md [hunk 11]
```diff
@@ -6,16 +6,36 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
 - The `rand-extension` example has been adapted to an updated version of the `ChainExtension` API ‒ [#1356](https://github.com/paritytech/ink/pull/1356)
 
-### Compatibility
+### Changed
+- :x: Contract size optimization in case contract doesn't accept payment ‒ [#1267](https://github.com/paritytech/ink/pull/1267) (thanks [@xgreenx](https://github.com/xgreenx)).
+- Move linting into ink repo - [#1361](https://github.com/paritytech/ink/pull/1267)
```
Comment (cmichi) [inline CHANGELOG.md:33]:
```suggestion
- Move ink! linter into `ink` repository ‒ [#1361](https://github.com/paritytech/ink/pull/1267)
```

File: CHANGELOG.md [hunk 12]
```diff
@@ -6,16 +6,36 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+
+### Removed
+- :x: Implement ecdsa_to_eth_address() and remove eth_compatibility crate ‒ [#1233](https://github.com/paritytech/ink/pull/1233)
 
+### Compatibility
```
Comment (cmichi) [inline CHANGELOG.md:38]: Can you make the compatibility section the first section under `## Version 4.0.0-alpha.1`? Then it'll be consistent with the other releases.

**Review by cmichi — APPROVED** [review]
Thank you! 🙏

**Comment by HCastano** [comment]
The ink-waterfall failure is related to some of the metadata changes that I want to address by publishing this release, so I'm gonna merge this anyways.

**Pull request merged**